ACPs/77-reinventing-subnets/README.md

```diff
@@ -46,77 +46,154 @@ For standalone networks like the Avalanche Primary Network, this is done by conn
+    // Must be Ed25519 NodeID
+    NodeID ids.NodeID `json:"nodeID"`
+    // Weight of this validator used when sampling
+    Weight uint64 `json:"weight"`
+    // [Signer] is the BLS key for this validator.
+    // Note: We do not enforce that the BLS key is unique across all validators.
+    //       This means that validators can share a key if they so choose.
+    //       However, a NodeID + Subnet does uniquely map to a BLS key
+    Signer signer.Signer `json:"signer"`
+    // Leftover $AVAX from the [Balance] will be issued to this
+    // owner once it is removed from the validator set.
+    ChangeOwner fx.Owner `json:"changeOwner"`
+}
+
+type RecoverSubnetTx struct {
+    // Metadata, inputs and outputs
+    BaseTx
+    // Validator set for the Subnet
+    Validators []SubnetValidator `json:"validators"`
+    // Auth that will be allowing these validators into the network
+    SubnetAuth verify.Verifiable `json:"subnetAuthorization"`
+}
+```
+
+### Adding Subnet Validators
+
+#### RegisterSubnetValidatorTx
+
+A `RegisterSubnetValidatorTx` can be used to add a Subnet Validator.
 
 ```golang
 type RegisterSubnetValidatorTx struct {
     // Metadata, inputs and outputs
     BaseTx
-    // Balance <= sum($AVAX inputs) - sum($AVAX outputs) - TxFee
+    // Balance must be equal to the balance uint64 in the [Message].
+    // Balance <= sum($AVAX inputs) - sum($AVAX outputs) - TxFee.
     Balance uint64 `json:"balance"`
     // [Signer] is the BLS key for this validator.
     // Note: We do not enforce that the BLS key is unique across all validators.
     //       This means that validators can share a key if they so choose.
     //       However, a NodeID does uniquely map to a BLS key
     Signer signer.Signer `json:"signer"`
-    // Leftover $AVAX from the [Balance] will be issued to this
-    // owner after it is removed from the validator set.
+    // Leftover $AVAX from the Subnet Validator's Balance will be issued to
+    // this owner after it is removed from the validator set.
     ChangeOwner fx.Owner `json:"changeOwner"`
-    // Warp message should include:
+    // AddressedCall with Payload:
     //   - SubnetID
     //   - NodeID (must be Ed25519 NodeID)
-    //   - Weight of the validator
+    //   - Weight
     //   - Expiry
     //   - Ed25519 Signature over "[subnetID]+[nodeID]+[blsPublicKey]+[weight]+[balance]+[expiry]"
-    //   - BLS multisig over the above payload
     Message warp.Message `json:"message"`
 }
 ```
 
-After the `RegisterSubnetValidatorTx` is accepted on the P-Chain, the Subnet Validator is added to the Subnet's validator set. The P-Chain will then send a Warp message to the Subnet notifying of the validator set addition.
+The `Message` field must be an `AddressedCall` with the payload:
+
+```text
++-----------+----------+-----------+
+|   codecID :   uint16 |   2 bytes |
++-----------+----------+-----------+
+|    typeID :   uint32 |   4 bytes |
++-----------+----------+-----------+
```

Comment:
These two fields are not represented in the type spec comment above.

Reply:
https://github.com/dhrubabasu/ACPs/pull/1#discussion_r1696100009

Reply:
IMO it should be reflected in the type spec to avoid confusion, otherwise we will probably get more questions about this

Reply:
By virtue of being an `AddressedCall` payload, there is a `codecID` and `typeID` (Ref: https://github.com/ava-labs/avalanchego/tree/master/vms/platformvm/warp/payload#addressedcall)

Reply:
The `codecID` and `typeID` for the `AddressedCall` payload of the Warp message are different than the `codecID` and `typeID` for the payload within the `AddressedCall` body though, no?

Reply:
Yes, that's correct. The `codecID` and `typeID` here are for the payload within the `AddressedCall`.